File: taurus-protect-sdk-java/client/src/main/java/com/taurushq/sdk/protect/client/cache/RulesContainerCache.java
package com.taurushq.sdk.protect.client.cache;

import com.taurushq.sdk.protect.client.model.ApiException;
import com.taurushq.sdk.protect.client.model.GovernanceRules;
import com.taurushq.sdk.protect.client.model.rulescontainer.DecodedRulesContainer;
import com.taurushq.sdk.protect.client.service.GovernanceRuleService;

import static com.google.common.base.Preconditions.checkArgument;
import static com.google.common.base.Preconditions.checkNotNull;

/**
 * Thread-safe cache for the decoded rules container with configurable TTL.
 * <p>
 * This cache stores the decoded rules container and refreshes it automatically
 * when the TTL expires. Used primarily for address signature verification
 * against HSM slot public keys.
 * <p>
 * This implementation avoids holding a lock during network I/O by using a
 * "fetching" flag pattern. When a refresh is needed, one thread will perform
 * the fetch while other threads wait for completion.
 */
public class RulesContainerCache {

    /**
     * Default cache TTL: 5 minutes.
     */
    public static final long DEFAULT_CACHE_TTL_MS = 5 * 60 * 1000L;

    private final GovernanceRuleService governanceRuleService;
    private final long cacheTtlMs;

    private final Object lock = new Object();
    private DecodedRulesContainer cachedContainer;
    private long cacheTimestamp;

    /**
     * Flag indicating whether a fetch is currently in progress.
     */
    private boolean fetching = false;

    /**
     * Exception captured during fetch, to be re-thrown to waiting threads.
     */
    private ApiException fetchException = null;

    /**
     * Creates a new rules container cache with default TTL (5 minutes).
     *
     * @param governanceRuleService the governance rule service for fetching rules
     */
    public RulesContainerCache(GovernanceRuleService governanceRuleService) {
        this(governanceRuleService, DEFAULT_CACHE_TTL_MS);
    }

    /**
     * Creates a new rules container cache with a custom TTL.
     *
     * @param governanceRuleService the governance rule service for fetching rules
     * @param cacheTtlMs            the cache time-to-live in milliseconds
     */
    public RulesContainerCache(GovernanceRuleService governanceRuleService, long cacheTtlMs) {
        checkNotNull(governanceRuleService, "governanceRuleService cannot be null");
        checkArgument(cacheTtlMs > 0, "cacheTtlMs must be positive");

        this.governanceRuleService = governanceRuleService;
        this.cacheTtlMs = cacheTtlMs;
    }

    /**
     * Gets the decoded rules container, fetching from the API if the cache is expired.
     * <p>
     * This method is thread-safe and will only fetch once if multiple threads
     * attempt to refresh simultaneously. The lock is released during network I/O
     * to prevent blocking other threads unnecessarily.
     *
     * @return the decoded rules container
     * @throws ApiException if fetching the rules fails
     */
    public DecodedRulesContainer getDecodedRulesContainer() throws ApiException {
        synchronized (lock) {
            long now = System.currentTimeMillis();

            // If cache is valid, return immediately
            if (cachedContainer != null && (now - cacheTimestamp) <= cacheTtlMs) {
                return cachedContainer;
            }

            // If another thread is already fetching, wait for it
            while (fetching) {
                try {
                    lock.wait();
                } catch (InterruptedException e) {
                    Thread.currentThread().interrupt();
                    ApiException ex = new ApiException();
                    ex.setCode(503);
                    ex.setError("Interrupted");
                    ex.setMessage("Interrupted while waiting for rules container refresh");
                    ex.initCause(e);
                    throw ex;
                }
            }

            // Re-check cache after waking up (another thread may have refreshed it)
            now = System.currentTimeMillis();
            if (cachedContainer != null && (now - cacheTimestamp) <= cacheTtlMs) {
                return cachedContainer;
            }

            // Check if the previous fetch failed
            if (fetchException != null) {
                ApiException ex = fetchException;
                fetchException = null;
                throw ex;
            }

            // This thread will perform the fetch
            fetching = true;
            fetchException = null;
        }

        // Perform network I/O outside the lock
        DecodedRulesContainer newContainer = null;
        ApiException exception = null;
        boolean settled = false;
        try {
            newContainer = doFetch();
            settled = true;
        } catch (ApiException e) {
            exception = e;
            settled = true;
        } finally {
            if (!settled) {
                releaseTheFetchSlot();
            }
        }

        // Update cache and notify waiting threads
        synchronized (lock) {
            fetching = false;
            if (exception != null) {
                fetchException = exception;
                lock.notifyAll();
                throw exception;
            }
            cachedContainer = newContainer;
            cacheTimestamp = System.currentTimeMillis();
            lock.notifyAll();
            return cachedContainer;
        }
    }

    /**
     * Clears the single-flight flag and wakes every waiter, for the path where the fetch
     * threw something the {@code catch (ApiException)} above does not see.
     *
     * <p>This is not defensive tidying. {@code doFetch()} runs governance verification,
     * whose {@code IntegrityException} is <b>unchecked</b> ({@code extends
     * SecurityException}), and the error-body path can raise {@code StackOverflowError}.
     * Either one escaped before {@code fetching} was cleared, so every later caller parked
     * on the untimed {@code lock.wait()} and never woke: <b>one crafted {@code /rules}
     * response permanently wedged address, asset and price verification process-wide</b> --
     * a denial of service on the whole verification surface from a single response.
     *
     * <p>{@code fetchException} is deliberately NOT set here: it is typed
     * {@code ApiException} and the escaping throwable propagates to this caller on its
     * own. Waiters re-check the cache, find it stale, and take a fresh turn at the fetch
     * -- which is the correct outcome, since nothing was cached.
     */
    private void releaseTheFetchSlot() {
        synchronized (lock) {
            fetching = false;
            lock.notifyAll();
        }
    }

    /**
     * Forces a cache refresh, fetching the latest rules from the API.
     * <p>
     * This method is thread-safe and releases the lock during network I/O.
     *
     * @throws ApiException if fetching the rules fails
     */
    public void invalidate() throws ApiException {
        synchronized (lock) {
            // Wait if another thread is already fetching
            while (fetching) {
                try {
                    lock.wait();
                } catch (InterruptedException e) {
                    Thread.currentThread().interrupt();
                    ApiException ex = new ApiException();
                    ex.setCode(503);
                    ex.setError("Interrupted");
                    ex.setMessage("Interrupted while waiting for rules container invalidation");
                    ex.initCause(e);
                    throw ex;
                }
            }

            // This thread will perform the fetch
            fetching = true;
            fetchException = null;
        }

        // Perform network I/O outside the lock
        DecodedRulesContainer newContainer = null;
        ApiException exception = null;
        boolean settled = false;
        try {
            newContainer = doFetch();
            settled = true;
        } catch (ApiException e) {
            exception = e;
            settled = true;
        } finally {
            if (!settled) {
                releaseTheFetchSlot();
            }
        }

        // Update cache and notify waiting threads
        synchronized (lock) {
            fetching = false;
            if (exception != null) {
                fetchException = exception;
                lock.notifyAll();
                throw exception;
            }
            cachedContainer = newContainer;
            cacheTimestamp = System.currentTimeMillis();
            lock.notifyAll();
        }
    }

    /**
     * Checks if the cache is currently valid (not expired).
     *
     * @return true if the cache is valid, false if expired or empty
     */
    public boolean isCacheValid() {
        synchronized (lock) {
            if (cachedContainer == null) {
                return false;
            }
            long now = System.currentTimeMillis();
            return (now - cacheTimestamp) <= cacheTtlMs;
        }
    }

    /**
     * Gets the configured cache TTL in milliseconds.
     *
     * @return the cache TTL
     */
    public long getCacheTtlMs() {
        return cacheTtlMs;
    }

    /**
     * Performs the actual fetch from the governance rule service.
     * This method should be called without holding the lock.
     *
     * @return the decoded rules container
     * @throws ApiException if fetching the rules fails
     */
    private DecodedRulesContainer doFetch() throws ApiException {
        GovernanceRules rules = governanceRuleService.getRules();
        if (rules == null) {
            ApiException ex = new ApiException();
            ex.setCode(503);  // Service Unavailable - rules service not responding
            ex.setError("ServiceUnavailable");
            ex.setMessage("No governance rules available from API");
            throw ex;
        }
        return governanceRuleService.getDecodedRulesContainer(rules);
    }
}

File: taurus-protect-sdk-python/taurus_protect/cache/rules_container_cache.py
"""Thread-safe cache for decoded governance rules container."""

from __future__ import annotations

import threading
import time
from typing import TYPE_CHECKING, Optional

from taurus_protect.errors import APIError
from taurus_protect.models.governance_rules import DecodedRulesContainer

if TYPE_CHECKING:
    from taurus_protect.services.governance_rule_service import GovernanceRuleService


class RulesContainerCache:
    """
    Thread-safe cache for the decoded rules container with configurable TTL.

    This cache stores the decoded rules container and refreshes it automatically
    when the TTL expires. Used primarily for address signature verification
    against HSM slot public keys.

    The implementation uses a "fetching" flag with Condition coordination to avoid
    holding the lock during network I/O, which could cause deadlocks under load.

    Example:
        >>> cache = RulesContainerCache(governance_service, ttl_ms=300_000)
        >>> rules = cache.get_decoded_rules_container()
        >>> hsm_key = rules.get_hsm_public_key()

    Attributes:
        DEFAULT_CACHE_TTL_MS: Default TTL of 5 minutes (300,000 ms).
    """

    DEFAULT_CACHE_TTL_MS: int = 5 * 60 * 1000  # 5 minutes

    def __init__(
        self,
        governance_rule_service: "GovernanceRuleService",
        ttl_ms: int = DEFAULT_CACHE_TTL_MS,
    ) -> None:
        """
        Create a new rules container cache.

        Args:
            governance_rule_service: Service for fetching governance rules.
            ttl_ms: Cache time-to-live in milliseconds. Defaults to 5 minutes.

        Raises:
            ValueError: If governance_rule_service is None or ttl_ms <= 0.
        """
        if governance_rule_service is None:
            raise ValueError("governance_rule_service cannot be None")
        if ttl_ms <= 0:
            raise ValueError("ttl_ms must be positive")

        self._governance_rule_service = governance_rule_service
        self._ttl_ms = ttl_ms
        self._lock = threading.RLock()
        self._condition = threading.Condition(self._lock)
        self._cached_container: Optional[DecodedRulesContainer] = None
        # Use monotonic time to be immune to system clock changes
        self._cache_timestamp_mono: float = 0.0
        self._fetching: bool = False

    @property
    def ttl_ms(self) -> int:
        """Get the configured cache TTL in milliseconds."""
        return self._ttl_ms

    def _is_cache_expired(self) -> bool:
        """Check if cache is expired (must be called under lock)."""
        if self._cached_container is None:
            return True
        elapsed_ms = (time.monotonic() - self._cache_timestamp_mono) * 1000
        return elapsed_ms > self._ttl_ms

    def get_decoded_rules_container(self) -> DecodedRulesContainer:
        """
        Get the decoded rules container, fetching from API if cache is expired.

        This method is thread-safe and will only fetch once if multiple threads
        attempt to refresh simultaneously. The lock is released during network I/O
        to prevent deadlocks under concurrent load.

        Returns:
            The decoded rules container.

        Raises:
            APIError: If fetching the rules fails.
        """
        with self._condition:
            # If cache is valid, return immediately
            if not self._is_cache_expired():
                return self._cached_container  # type: ignore[return-value]

            # If another thread is already fetching, wait for it
            while self._fetching:
                self._condition.wait()
                # After waking, check if cache is now valid
                if not self._is_cache_expired():
                    return self._cached_container  # type: ignore[return-value]

            # We need to fetch - mark as fetching
            self._fetching = True

        # Network I/O happens OUTSIDE the lock to prevent deadlocks
        try:
            new_container = self._fetch_rules_container()
        finally:
            with self._condition:
                self._fetching = False
                self._condition.notify_all()

        # Update cache under lock
        with self._condition:
            self._cached_container = new_container
            self._cache_timestamp_mono = time.monotonic()
            return new_container

    def invalidate(self) -> None:
        """
        Force a cache refresh, fetching the latest rules from the API.

        Raises:
            APIError: If fetching the rules fails.
        """
        with self._condition:
            # If another thread is already fetching, wait for it
            while self._fetching:
                self._condition.wait()

            # Mark as fetching
            self._fetching = True

        # Network I/O happens OUTSIDE the lock
        try:
            new_container = self._fetch_rules_container()
        finally:
            with self._condition:
                self._fetching = False
                self._condition.notify_all()

        # Update cache under lock
        with self._condition:
            self._cached_container = new_container
            self._cache_timestamp_mono = time.monotonic()

    def is_cache_valid(self) -> bool:
        """
        Check if the cache is currently valid (not expired).

        Returns:
            True if the cache is valid, False if expired or empty.
        """
        with self._lock:
            return not self._is_cache_expired()

    def clear(self) -> None:
        """Clear the cache without refreshing."""
        with self._lock:
            self._cached_container = None
            self._cache_timestamp_mono = 0.0

    def _fetch_rules_container(self) -> DecodedRulesContainer:
        """
        Fetch governance rules from the API.

        This method performs network I/O and should be called WITHOUT holding the lock.

        Raises:
            APIError: If fetching fails or no rules are returned.
        """
        rules = self._governance_rule_service.get_rules()
        if rules is None:
            raise APIError("No governance rules available from API")

        # Get the decoded rules container (with signature verification)
        container = self._governance_rule_service.get_decoded_rules_container(rules)
        if container is None:
            raise APIError("Failed to decode governance rules container")
        return container
